ext/openssl/openssl.c

```diff
@@ -5309,6 +5330,717 @@ PHP_FUNCTION(openssl_pkcs7_decrypt)
 
 /* }}} */
 
+/* {{{ CMS S/MIME functions taken from PKCS#7 functions */
+
+/* {{{ proto bool openssl_cms_verify(string filename, int flags [, string signerscerts [, array cainfo [, string extracerts [, string content [, string pk7]]]], string sigfile, int encoding])
+   Verifies that the data block is intact, the signer is who they say they are, and returns the CERTs of the signers */
+PHP_FUNCTION(openssl_cms_verify)
+{
+	X509_STORE * store = NULL;
+	zval * cainfo = NULL;
+	STACK_OF(X509) *signers= NULL;
+	STACK_OF(X509) *others = NULL;
+	CMS_ContentInfo * cms = NULL;
+	BIO * in = NULL, * datain = NULL, * dataout = NULL, * p7bout  = NULL;
+	BIO *certout = NULL, *sigbio = NULL;
+	zend_long flags = 0;
+	char * filename;
```

Comment:
Save yourself some pain and use `zend_string*` for your strings.

```
zend_string* arg;

if (zend_parse_parameters(ZEND_NUM_ARGS(), "S", &arg) == FAILURE) { return; }
```

Or in your case since you don't want null bytes, use the "P" format character.

Then when accessing the string values, `ZSTR_VAL(arg)` and `ZSTR_LEN(arg)` to get at their component pieces.

Additionally, having these as `zend_string` values may help later when calling hash/symtable functions or other engine internals.


You can also declare these in a single (or handful) line to save vertical space:

```
zend_string *filename, *extracerts = NULL, *signersfilename = NULL, *datafilename = NULL;
zend_string *p7bfilename = NULL, *sigfile = NULL;
```

Note that I didn't initialize `filename` since it's a required arg, and thus you'll either get it, or your function will exit.

Reply:
I agree that's better to use `zend_string` if the value is used later for hash lookup or something PHP specific. But in this case it's used just once for `BIO_new_file`. There is really no point to use `zend_string`. It will just require extra macro around that which is not really useful.

Reply:
And it's also used in `php_openssl_open_base_dir_chk` so it's more consistent to keep it as `char *`